File: csharp/AppEncryption/AppEncryption/Persistence/InMemoryMetastoreImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using LanguageExt;

namespace GoDaddy.Asherah.AppEncryption.Persistence
{
    /// <summary>
    /// Provides a volatile implementation of <see cref="IMetastore{T}"/> for values using a
    /// <see cref="System.Data.DataTable"/>. NOTE: This should NEVER be used in a production environment.
    /// </summary>
    ///
    /// <typeparam name="T">The type of value to store and retrieve.</typeparam>
    public class InMemoryMetastoreImpl<T> : IMetastore<T>
    {
        private readonly DataTable dataTable;

        /// <summary>
        /// Initializes a new instance of the <see cref="InMemoryMetastoreImpl{T}"/> class, with 3 columns.
        /// <code>
        /// keyId | created | value
        /// ----- | ------- | ------
        ///       |         |
        ///       |         |
        /// </code>
        /// Uses 'keyId' and 'created' as the primary key.
        /// </summary>
        public InMemoryMetastoreImpl()
        {
            dataTable = new DataTable();
            dataTable.Columns.Add("keyId", typeof(string));
            dataTable.Columns.Add("created", typeof(DateTimeOffset));
            dataTable.Columns.Add("value", typeof(T));
            dataTable.PrimaryKey = new[] { dataTable.Columns["keyId"], dataTable.Columns["created"] };
        }

        /// <inheritdoc />
        public virtual Option<T> Load(string keyId, DateTimeOffset created)
        {
            lock (dataTable)
            {
                List<DataRow> dataRows = dataTable.Rows.Cast<DataRow>()
                    .Where(row => row["keyId"].Equals(keyId)
                                  && row["created"].Equals(created))
                    .ToList();
                if (!dataRows.Any())
                {
                    return Option<T>.None;
                }

                return (T)dataRows.Single()["value"];
            }
        }

        /// <summary>
        /// Obtain the latest value associated with the keyId by ordering the datatable in ascending order.
        /// </summary>
        ///
        /// <param name="keyId">The keyId to lookup.</param>
        /// <returns>The latest <see cref="T"/> value associated with the keyId, if any.</returns>
        public virtual Option<T> LoadLatest(string keyId)
        {
            lock (dataTable)
            {
                List<DataRow> dataRows = dataTable.Rows.Cast<DataRow>()
                    .Where(row => row["keyId"].Equals(keyId))
                    .OrderBy(row => row["created"])
                    .ToList();

                // Need to check if empty as Last will throw an exception instead of returning null
                if (!dataRows.Any())
                {
                    return Option<T>.None;
                }

                return (T)dataRows.Last()["value"];
            }
        }

        /// <inheritdoc />
        public virtual bool Store(string keyId, DateTimeOffset created, T value)
        {
            lock (dataTable)
            {
                List<DataRow> dataRows = dataTable.Rows.Cast<DataRow>()
                    .Where(row => row["keyId"].Equals(keyId)
                                  && row["created"].Equals(created))
                    .ToList();
                if (dataRows.Any())
                {
                    return false;
                }

                dataTable.Rows.Add(keyId, created, value);
                return true;
            }
        }
    }
}

File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/persistence/InMemoryMetastoreImpl.java
package com.godaddy.asherah.appencryption.persistence;

import java.time.Instant;
import java.util.Optional;
import java.util.SortedMap;

import com.google.common.collect.TreeBasedTable;

/**
 * Provides a {@link TreeBasedTable} based implementation of {@link Metastore} to store and retrieve
 * {@link com.godaddy.asherah.appencryption.utils.Json} values for system and intermediate keys.
 * NOTE: This is a volatile implementation and should NEVER be used in a production environment.
 *
 * @param <T> The type of the value being loaded and stored in the metastore.
 */
public class InMemoryMetastoreImpl<T> implements Metastore<T> {
  private final TreeBasedTable<String, Instant, T> table = TreeBasedTable.create();

  @Override
  public Optional<T> load(final String keyId, final Instant created) {
    synchronized (table) {
      return Optional.ofNullable(table.get(keyId, created));
    }
  }

  @Override
  public Optional<T> loadLatest(final String keyId) {
    synchronized (table) {
      SortedMap<Instant, T> partitionMap = table.row(keyId);
      // Need to check if empty as lastKey will throw an exception instead of returning null
      if (partitionMap.isEmpty()) {
        return Optional.empty();
      }

      return Optional.of(partitionMap.get(partitionMap.lastKey()));
    }
  }

  @Override
  public boolean store(final String keyId, final Instant created, final T value) {
    synchronized (table) {

      if (table.contains(keyId, created)) {
        return false;
      }

      table.put(keyId, created, value);
      return true;
    }
  }
}
